File: test/genome/test_property_genome.py
"""test_property_genome.py — Property-based genome codec tests (T9, PKT-P1).

Uses Hypothesis to verify:
1. decode(encode(x)) == x (round-trip)
2. Any single-char change to the mutable body fails checksum
3. Every ±1 mutation yields either a valid genome or an explicit typed error

Runtime target: < 30 seconds in CI.
"""

from __future__ import annotations

import pytest
from hypothesis import given, settings, HealthCheck
from hypothesis import strategies as st

from alienclaw.genome.alphabet import ALPHABET, GENOME_LENGTH, SECTION_LENGTH
from alienclaw.genome.codec import assemble, parse, round_trip_check
from alienclaw.genome.checksum import compute_checksum, verify_checksum

ALPHABET_STR = ALPHABET

# ---------------------------------------------------------------------------
# Strategies
# ---------------------------------------------------------------------------

@st.composite
def section64(draw: st.DrawFn) -> str:
    return "".join(draw(st.sampled_from(ALPHABET_STR)) for _ in range(SECTION_LENGTH))


@st.composite
def valid_genome(draw: st.DrawFn) -> str:
    identity = draw(section64())
    execution = draw(section64())
    behavior = draw(section64())
    return assemble(identity, execution, behavior)


# ---------------------------------------------------------------------------
# Property 1: round-trip
# ---------------------------------------------------------------------------

@given(section64(), section64(), section64())
@settings(max_examples=200, deadline=None, suppress_health_check=[HealthCheck.too_slow])
def test_roundtrip_encode_decode(identity: str, execution: str, behavior: str) -> None:
    """assemble then parse should recover the original three sections."""
    genome = assemble(identity, execution, behavior)
    parsed = parse(genome)
    assert parsed.identity  == identity
    assert parsed.execution == execution
    assert parsed.behavior  == behavior


@given(valid_genome())
@settings(max_examples=200, deadline=None)
def test_roundtrip_full(genome: str) -> None:
    """round_trip_check must be True for any assembled genome."""
    assert round_trip_check(genome)


# ---------------------------------------------------------------------------
# Property 2: single-char body mutation fails checksum
# ---------------------------------------------------------------------------

@given(valid_genome(), st.integers(min_value=0, max_value=191))
@settings(max_examples=200, deadline=None)
def test_body_mutation_invalidates_checksum(genome: str, position: int) -> None:
    """Flipping any char in positions 0..191 must produce a checksum mismatch."""
    idx = ALPHABET_STR.index(genome[position])
    next_char = ALPHABET_STR[(idx + 1) % len(ALPHABET_STR)]
    mutated = genome[:position] + next_char + genome[position + 1:]
    assert len(mutated) == GENOME_LENGTH
    # The checksum must no longer verify
    assert not verify_checksum(mutated), (
        f"Mutation at position {position} did not invalidate checksum\n"
        f"original={genome!r}\nmutated={mutated!r}"
    )


# ---------------------------------------------------------------------------
# Property 3: ±1 mutation yields valid genome OR explicit ValueError
# ---------------------------------------------------------------------------

@given(valid_genome(), st.integers(min_value=0, max_value=GENOME_LENGTH - 1))
@settings(max_examples=200, deadline=None)
def test_mutation_yields_valid_or_explicit_error(genome: str, position: int) -> None:
    """Every ±1 mutation either produces a valid genome or raises ValueError."""
    idx = ALPHABET_STR.index(genome[position])
    next_char = ALPHABET_STR[(idx + 1) % len(ALPHABET_STR)]
    mutated = genome[:position] + next_char + genome[position + 1:]

    try:
        parsed = parse(mutated)
        # If parse succeeds, the genome must actually have a valid checksum
        assert verify_checksum(mutated)
        # And parse must be deterministic
        parsed2 = parse(mutated)
        assert parsed.identity  == parsed2.identity
        assert parsed.execution == parsed2.execution
        assert parsed.behavior  == parsed2.behavior
    except ValueError:
        # Explicit error is acceptable
        pass


# ---------------------------------------------------------------------------
# Property: non-Base62 chars are always rejected
# ---------------------------------------------------------------------------

NON_BASE62 = ["!", "@", " ", "\n", "\0", "+", "/", "="]


@given(valid_genome(), st.integers(min_value=0, max_value=GENOME_LENGTH - 1),
       st.sampled_from(NON_BASE62))
@settings(max_examples=200, deadline=None)
def test_non_base62_always_rejected(genome: str, position: int, bad_char: str) -> None:
    """Inserting any non-Base62 character must cause parse() to raise ValueError."""
    mutated = genome[:position] + bad_char + genome[position + 1:]
    with pytest.raises(ValueError):
        parse(mutated)


# ---------------------------------------------------------------------------
# Property: wrong-length inputs always rejected
# ---------------------------------------------------------------------------

@given(st.text(alphabet=ALPHABET_STR, min_size=0, max_size=255))
@settings(max_examples=100, deadline=None)
def test_short_genome_rejected(short: str) -> None:
    with pytest.raises(ValueError, match="256"):
        parse(short)


@given(st.text(alphabet=ALPHABET_STR, min_size=257, max_size=300))
@settings(max_examples=100, deadline=None)
def test_long_genome_rejected(long: str) -> None:
    with pytest.raises(ValueError, match="256"):
        parse(long)


# ---------------------------------------------------------------------------
# Cross-language consistency note
#
# Both TS and Python codecs are tested against the shared fixture file at
# test/fixtures/genome-spec-fixtures.json. The property tests above verify
# the Python codec's invariants. For cross-language equivalence, the fixture
# runner tests (test/genome/ts-fixture-runner.test.ts + test_fixtures.py)
# assert that both codecs produce identical results for every fixture case.
# ---------------------------------------------------------------------------

File: test/genome/test_property_genome.ts
/**
 * test/genome/test_property_genome.ts — Property-based genome codec tests (T9, PKT-P1).
 *
 * Uses fast-check to generate random inputs and verify invariants:
 * 1. decode(encode(x)) === x (round-trip)
 * 2. Any single-char change to the mutable body fails checksum validation
 * 3. Every ±1 char mutation yields either a valid genome or an explicit typed error
 *
 * Runtime target: < 30 seconds in CI.
 */

import { describe, it, expect } from 'vitest';
import * as fc from 'fast-check';
import {
  assembleGenome,
  parseGenome,
  validateGenome,
  computeChecksum,
  BASE62_ALPHABET,
  GENOME_LENGTH,
  SECTION_SIZE,
} from '../../src/alienclaw/registry/genome-codec.js';

// ---------------------------------------------------------------------------
// Arbitrary: random 64-char Base62 section
// ---------------------------------------------------------------------------

const base62Char = fc.constantFrom(...BASE62_ALPHABET.split(''));
const section64 = fc.array(base62Char, { minLength: 64, maxLength: 64 })
  .map((chars) => chars.join(''));

// ---------------------------------------------------------------------------
// Arbitrary: random valid 256-char genome
// ---------------------------------------------------------------------------

const validGenome = fc
  .tuple(section64, section64, section64)
  .map(([identity, execution, behavior]) => assembleGenome(identity, execution, behavior));

// ---------------------------------------------------------------------------
// Property 1: decode(encode(x)) === x (round-trip)
// ---------------------------------------------------------------------------

describe('Property: round-trip encode/decode', () => {
  it('parseGenome(assembleGenome(i, e, b)) reproduces all three sections', () => {
    fc.assert(
      fc.property(section64, section64, section64, (identity, execution, behavior) => {
        const genome = assembleGenome(identity, execution, behavior);
        const parsed = parseGenome(genome);
        return (
          parsed.identity  === identity  &&
          parsed.execution === execution &&
          parsed.behavior  === behavior
        );
      }),
      { numRuns: 200, seed: 42 },
    );
  });

  it('the checksum section is exactly 64 chars and all Base62', () => {
    fc.assert(
      fc.property(section64, section64, section64, (identity, execution, behavior) => {
        const genome = assembleGenome(identity, execution, behavior);
        const checksum = genome.slice(SECTION_SIZE * 3);
        if (checksum.length !== SECTION_SIZE) return false;
        return [...checksum].every((c) => BASE62_ALPHABET.includes(c));
      }),
      { numRuns: 200, seed: 43 },
    );
  });
});

// ---------------------------------------------------------------------------
// Property 2: Any single-char change to the body fails checksum
// ---------------------------------------------------------------------------

describe('Property: single-char body mutation invalidates checksum', () => {
  it('flipping any char in positions 0..191 fails checksum validation', () => {
    fc.assert(
      fc.property(
        validGenome,
        fc.integer({ min: 0, max: 191 }),
        (genome, position) => {
          // Flip the char to the next Base62 char
          const idx = BASE62_ALPHABET.indexOf(genome[position]!);
          const nextChar = BASE62_ALPHABET[(idx + 1) % BASE62_ALPHABET.length]!;
          const mutated = genome.slice(0, position) + nextChar + genome.slice(position + 1);
          const result = validateGenome(mutated);
          // Must be invalid — specifically checksum mismatch
          return !result.valid && result.errors.some((e) => e.includes('hecksum'));
        },
      ),
      { numRuns: 200, seed: 44 },
    );
  });
});

// ---------------------------------------------------------------------------
// Property 3: Every ±1 mutation yields valid genome OR explicit typed error
// ---------------------------------------------------------------------------

describe('Property: ±1 mutation yields valid or explicit error', () => {
  it('mutating any single position produces valid genome or explicit error', () => {
    fc.assert(
      fc.property(
        validGenome,
        fc.integer({ min: 0, max: GENOME_LENGTH - 1 }),
        (genome, position) => {
          const idx = BASE62_ALPHABET.indexOf(genome[position]!);
          const nextChar = BASE62_ALPHABET[(idx + 1) % BASE62_ALPHABET.length]!;
          const mutated = genome.slice(0, position) + nextChar + genome.slice(position + 1);

          // Attempt to validate the mutated genome
          try {
            const result = validateGenome(mutated);
            // Must either be valid or have explicit errors (never silently wrong)
            if (result.valid) {
              // Valid mutation: checksum must match
              const body = mutated.slice(0, SECTION_SIZE * 3);
              const stored = mutated.slice(SECTION_SIZE * 3);
              return computeChecksum(body) === stored;
            }
            // Invalid: must have at least one error message
            return result.errors.length > 0;
          } catch (err) {
            // Explicit thrown error is also acceptable
            return err instanceof Error && err.message.length > 0;
          }
        },
      ),
      { numRuns: 200, seed: 45 },
    );
  });
});

// ---------------------------------------------------------------------------
// Property: invalid length always rejected
// ---------------------------------------------------------------------------

describe('Property: wrong-length inputs always rejected', () => {
  it('genomes shorter than 256 are always invalid', () => {
    fc.assert(
      fc.property(
        fc.array(base62Char, { minLength: 0, maxLength: 255 }).map((cs) => cs.join('')),
        (shortGenome) => {
          const result = validateGenome(shortGenome);
          return !result.valid;
        },
      ),
      { numRuns: 100, seed: 46 },
    );
  });

  it('genomes longer than 256 are always invalid', () => {
    fc.assert(
      fc.property(
        fc.array(base62Char, { minLength: 257, maxLength: 300 }).map((cs) => cs.join('')),
        (longGenome) => {
          const result = validateGenome(longGenome);
          return !result.valid;
        },
      ),
      { numRuns: 100, seed: 47 },
    );
  });
});

// ---------------------------------------------------------------------------
// Property: non-Base62 chars always rejected
// ---------------------------------------------------------------------------

describe('Property: non-Base62 inputs always rejected', () => {
  it('inserting a non-Base62 char at any position fails validation', () => {
    const nonBase62 = ['!', '@', '#', '+', '/', '=', ' ', '\n', '\0'];
    fc.assert(
      fc.property(
        validGenome,
        fc.integer({ min: 0, max: GENOME_LENGTH - 1 }),
        fc.constantFrom(...nonBase62),
        (genome, position, badChar) => {
          const mutated = genome.slice(0, position) + badChar + genome.slice(position + 1);
          const result = validateGenome(mutated);
          return !result.valid;
        },
      ),
      { numRuns: 200, seed: 48 },
    );
  });
});
